File: lib/flash_profile/atom.ex
defmodule FlashProfile.Atom do
  @moduledoc """
  Implementation of atomic patterns (atoms) for FlashProfile.

  An atom α: String → Int is a function that matches a prefix of a string
  and returns the length matched (0 = no match). Atoms only match non-empty
  prefixes.

  From the paper (Definition 4.1):
  "An atom α: String → Int is a function, which given a string s, returns
  the length of the longest prefix of s that satisfies its constraints.
  Atoms only match non-empty prefixes. α(s) = 0 indicates match failure."

  ## Four Types of Atoms

  1. **Constant Strings (Const_s)**: Matches only the string s as prefix
  2. **Regular Expressions (RegEx_r)**: Returns length of longest prefix matched by regex r
  3. **Character Classes (Class^z_c)**:
     - Class^0_c: Returns length of longest prefix containing only chars from set c
     - Class^z_c (z > 0): Fixed-width variant - matches exactly z characters from set c
  4. **Arbitrary Functions (Funct_f)**: Custom matching function f
  """

  @type atom_type :: :constant | :char_class | :regex | :function

  @type t :: %__MODULE__{
          name: String.t(),
          type: atom_type(),
          matcher: (String.t() -> non_neg_integer()),
          static_cost: float(),
          params: map()
        }

  defstruct [
    :name,
    :type,
    :matcher,
    :static_cost,
    :params
  ]

  ## Constructor Functions

  @doc """
  Create a constant string atom.

  Matches exactly `string` as a prefix and returns the length of the string
  if matched, 0 otherwise.

  Cost is proportional to 1/length(string) to prefer longer matches.

  ## Examples

      iex> atom = FlashProfile.Atom.constant("PMC")
      iex> FlashProfile.Atom.match(atom, "PMC12345")
      3
      iex> FlashProfile.Atom.match(atom, "XYZ")
      0
  """
  @spec constant(String.t()) :: t()
  def constant(string) when is_binary(string) and byte_size(string) > 0 do
    len = String.length(string)
    # Cost proportional to 1/length - shorter strings have higher cost
    cost = 100.0 / len

    matcher = fn s ->
      if String.starts_with?(s, string) do
        len
      else
        0
      end
    end

    %__MODULE__{
      name: inspect(string),
      type: :constant,
      matcher: matcher,
      static_cost: cost,
      params: %{string: string, length: len}
    }
  end

  @doc """
  Create a variable-width character class atom.

  Matches the longest prefix where all characters are in the allowed set.
  Returns 0 if the first character doesn't match.

  ## Parameters

  - `name`: Display name for the atom (e.g., "Digit", "Upper")
  - `chars`: Charlist of allowed characters (e.g., ~c"0123456789")
  - `static_cost`: Base cost for this atom

  ## Examples

      iex> digit = FlashProfile.Atom.char_class("Digit", ~c"0123456789", 8.2)
      iex> FlashProfile.Atom.match(digit, "123abc")
      3
      iex> FlashProfile.Atom.match(digit, "abc123")
      0
  """
  @spec char_class(String.t(), charlist(), float()) :: t()
  def char_class(name, chars, static_cost) when is_list(chars) and is_float(static_cost) do
    # Convert charlist to MapSet for O(1) lookup
    char_set = MapSet.new(chars)

    matcher = fn s ->
      match_char_class_variable(s, char_set)
    end

    %__MODULE__{
      name: name,
      type: :char_class,
      matcher: matcher,
      static_cost: static_cost,
      params: %{chars: chars, width: 0, char_set: char_set}
    }
  end

  @doc """
  Create a fixed-width character class atom.

  Matches exactly `width` characters from the character set.
  Returns `width` if exactly `width` characters match, 0 otherwise.

  Cost is base_cost / width (fixed-width costs less than variable).

  ## Parameters

  - `name`: Display name for the atom
  - `chars`: Charlist of allowed characters
  - `width`: Exact number of characters to match (must be > 0)
  - `static_cost`: Base cost for this atom (will be divided by width)

  ## Examples

      iex> digit2 = FlashProfile.Atom.char_class("Digit", ~c"0123456789", 2, 8.2)
      iex> FlashProfile.Atom.match(digit2, "12345")
      2
      iex> FlashProfile.Atom.match(digit2, "1abc")
      0
  """
  @spec char_class(String.t(), charlist(), pos_integer(), float()) :: t()
  def char_class(name, chars, width, static_cost)
      when is_list(chars) and is_integer(width) and width > 0 and is_float(static_cost) do
    char_set = MapSet.new(chars)
    # Fixed-width cost is base_cost / width
    adjusted_cost = static_cost / width

    matcher = fn s ->
      match_char_class_fixed(s, char_set, width)
    end

    %__MODULE__{
      name: name,
      type: :char_class,
      matcher: matcher,
      static_cost: adjusted_cost,
      params: %{chars: chars, width: width, char_set: char_set}
    }
  end

  @doc """
  Create a regex atom.

  The pattern must match from the start of the string (should be anchored with ^).
  Returns the length of the matched prefix.

  ## Parameters

  - `name`: Display name for the atom
  - `pattern`: Regex pattern (string or compiled regex)
  - `static_cost`: Base cost for this atom

  ## Examples

      iex> atom = FlashProfile.Atom.regex("Email", ~r/^[a-z]+@/, 15.0)
      iex> FlashProfile.Atom.match(atom, "user@example.com")
      5
  """
  @spec regex(String.t(), String.t() | Regex.t(), float()) :: t()
  def regex(name, pattern, static_cost) when is_float(static_cost) do
    compiled_regex =
      case pattern do
        %Regex{} = r -> r
        string when is_binary(string) -> Regex.compile!(string)
      end

    matcher = fn s ->
      case Regex.run(compiled_regex, s, return: :index) do
        [{0, length}] -> length
        _ -> 0
      end
    end

    %__MODULE__{
      name: name,
      type: :regex,
      matcher: matcher,
      static_cost: static_cost,
      params: %{pattern: compiled_regex}
    }
  end

  @doc """
  Create a function atom with custom matching logic.

  The matcher function should take a string and return the length of the
  matched prefix (0 for no match).

  ## Parameters

  - `name`: Display name for the atom
  - `matcher_fn`: Function (String.t() -> non_neg_integer())
  - `static_cost`: Base cost for this atom

  ## Examples

      iex> matcher = fn s ->
      ...>   cond do
      ...>     String.starts_with?(s, "http://") -> 7
      ...>     String.starts_with?(s, "https://") -> 8
      ...>     true -> 0
      ...>   end
      ...> end
      iex> atom = FlashProfile.Atom.function("Protocol", matcher, 10.0)
      iex> FlashProfile.Atom.match(atom, "https://example.com")
      8
  """
  @spec function(String.t(), (String.t() -> non_neg_integer()), float()) :: t()
  def function(name, matcher_fn, static_cost)
      when is_function(matcher_fn, 1) and is_float(static_cost) do
    %__MODULE__{
      name: name,
      type: :function,
      matcher: matcher_fn,
      static_cost: static_cost,
      params: %{}
    }
  end

  ## Core Functions

  @doc """
  Match atom against string, return length of matched prefix.

  Returns 0 if no match (atoms only match non-empty prefixes).

  ## Examples

      iex> digit = FlashProfile.Atom.char_class("Digit", ~c"0123456789", 8.2)
      iex> FlashProfile.Atom.match(digit, "123abc")
      3
  """
  @spec match(t(), String.t()) :: non_neg_integer()
  def match(%__MODULE__{matcher: matcher}, string) when is_binary(string) do
    matcher.(string)
  end

  @doc """
  Get the static cost of an atom.

  ## Examples

      iex> atom = FlashProfile.Atom.constant("PMC")
      iex> FlashProfile.Atom.static_cost(atom)
      33.333333333333336
  """
  @spec static_cost(t()) :: float()
  def static_cost(%__MODULE__{static_cost: cost}), do: cost

  @doc """
  Format atom for display.

  Returns a string representation suitable for pattern printing.

  ## Examples

      iex> digit = FlashProfile.Atom.char_class("Digit", ~c"0123456789", 8.2)
      iex> FlashProfile.Atom.to_string(digit)
      "Digit"

      iex> const = FlashProfile.Atom.constant("PMC")
      iex> FlashProfile.Atom.to_string(const)
      "\\"PMC\\""
  """
  @spec to_string(t()) :: String.t()
  def to_string(%__MODULE__{name: name, type: type}) do
    case type do
      :constant -> name
      _ -> name
    end
  end

  @doc """
  Check if atom matches entire string.

  Returns true if the atom matches and consumes the entire string.

  ## Examples

      iex> digit = FlashProfile.Atom.char_class("Digit", ~c"0123456789", 8.2)
      iex> FlashProfile.Atom.matches_entirely?(digit, "123")
      true
      iex> FlashProfile.Atom.matches_entirely?(digit, "123abc")
      false
  """
  @spec matches_entirely?(t(), String.t()) :: boolean()
  def matches_entirely?(%__MODULE__{} = atom, string) when is_binary(string) do
    matched_length = match(atom, string)
    matched_length > 0 and matched_length == String.length(string)
  end

  @doc """
  Create a fixed-width variant of a character class atom.

  Given a variable-width char_class atom, creates a new atom that matches
  exactly `width` characters from the same character set.

  ## Examples

      iex> digit = FlashProfile.Atom.char_class("Digit", ~c"0123456789", 8.2)
      iex> digit2 = FlashProfile.Atom.with_fixed_width(digit, 2)
      iex> FlashProfile.Atom.match(digit2, "12345")
      2
      iex> FlashProfile.Atom.match(digit2, "1abc")
      0
  """
  @spec with_fixed_width(t(), pos_integer()) :: t()
  def with_fixed_width(
        %__MODULE__{type: :char_class, name: name, params: params, static_cost: base_cost},
        width
      )
      when is_integer(width) and width > 0 do
    chars = params.chars
    char_class(name, chars, width, base_cost)
  end

  ## Private Helper Functions

  # Match variable-width character class
  # Returns length of longest prefix where all characters are in char_set
  defp match_char_class_variable(string, char_set) do
    string
    |> String.graphemes()
    |> Enum.reduce_while(0, fn grapheme, count ->
      # Convert grapheme to codepoint for comparison
      codepoint = String.to_charlist(grapheme) |> List.first()

      if codepoint && MapSet.member?(char_set, codepoint) do
        {:cont, count + 1}
      else
        {:halt, count}
      end
    end)
  end

  # Match fixed-width character class
  # Returns width if exactly width characters match, 0 otherwise
  defp match_char_class_fixed(string, char_set, width) do
    graphemes = String.graphemes(string)

    if length(graphemes) < width do
      0
    else
      matched =
        graphemes
        |> Enum.take(width)
        |> Enum.all?(fn grapheme ->
          codepoint = String.to_charlist(grapheme) |> List.first()
          codepoint && MapSet.member?(char_set, codepoint)
        end)

      if matched, do: width, else: 0
    end
  end
end

File: native/flash_profile/atom.zig
const std = @import("std");
const types = @import("types.zig");

/// Efficient ASCII character set using bitmap for O(1) membership testing.
/// Uses two u64s to cover all 128 ASCII values (0-127).
pub const CharSet = struct {
    /// Lower 64 bits (ASCII 0-63)
    low: u64,
    /// Upper 64 bits (ASCII 64-127)
    high: u64,

    /// Create an empty character set.
    pub fn empty() CharSet {
        return .{ .low = 0, .high = 0 };
    }

    /// Create a character set from a slice of characters.
    pub fn fromChars(chars: []const u8) CharSet {
        var set = empty();
        for (chars) |c| {
            set.add(c);
        }
        return set;
    }

    /// Add a character to the set.
    pub fn add(self: *CharSet, c: u8) void {
        if (c < 64) {
            self.low |= @as(u64, 1) << @intCast(c);
        } else if (c < 128) {
            self.high |= @as(u64, 1) << @intCast(c - 64);
        }
        // Non-ASCII characters are ignored
    }

    /// Check if a character is in the set. O(1) lookup.
    pub fn contains(self: CharSet, c: u8) bool {
        if (c < 64) {
            return (self.low & (@as(u64, 1) << @intCast(c))) != 0;
        } else if (c < 128) {
            return (self.high & (@as(u64, 1) << @intCast(c - 64))) != 0;
        }
        return false;
    }
};

/// Type of atom for pattern matching.
pub const AtomType = enum {
    /// Constant string match
    constant,
    /// Character class (fixed or variable width)
    char_class,
    /// Regular expression match
    regex,
    /// Custom function match
    function,
};

/// Character class data for both fixed and variable width matching.
pub const CharClassData = struct {
    /// Set of allowed characters
    char_set: CharSet,
    /// Width: 0 = variable (greedy), >0 = fixed width
    width: u32,
};

/// Constant string data.
pub const ConstantData = struct {
    /// The exact string to match
    string: []const u8,
};

/// Atom data - type-specific matching information.
pub const AtomData = union(AtomType) {
    constant: ConstantData,
    char_class: CharClassData,
    regex: void, // Not implemented in Phase 2
    function: void, // Not implemented in Phase 2
};

/// An atom represents an atomic pattern that matches a prefix of a string.
///
/// From the FlashProfile paper (Definition 4.1):
/// "An atom α: String → Int is a function, which given a string s, returns
/// the length of the longest prefix of s that satisfies its constraints.
/// Atoms only match non-empty prefixes. α(s) = 0 indicates match failure."
pub const Atom = struct {
    /// Display name of the atom
    name: []const u8,

    /// Static cost for this atom (used in pattern selection)
    static_cost: f64,

    /// Type-specific matching data
    data: AtomData,

    /// Match this atom against a string.
    /// Returns the length of the matched prefix, or null if no match.
    ///
    /// Matching semantics:
    /// - Constant: Returns string length if prefix matches exactly, else null
    /// - CharClass (variable): Returns length of longest prefix with all chars in set
    /// - CharClass (fixed): Returns width if exactly width chars match, else null
    pub fn match(self: Atom, string: []const u8) ?usize {
        return switch (self.data) {
            .constant => |data| self.matchConstant(string, data),
            .char_class => |data| self.matchCharClass(string, data),
            .regex => null, // Not implemented
            .function => null, // Not implemented
        };
    }

    /// Match a constant string.
    fn matchConstant(self: Atom, string: []const u8, data: ConstantData) ?usize {
        _ = self;
        if (string.len < data.string.len) return null;

        // Check if string starts with the constant
        if (std.mem.eql(u8, string[0..data.string.len], data.string)) {
            return data.string.len;
        }
        return null;
    }

    /// Match a character class (variable or fixed width).
    fn matchCharClass(self: Atom, string: []const u8, data: CharClassData) ?usize {
        _ = self;
        if (data.width == 0) {
            // Variable width: match as many characters as possible
            return matchCharClassVariable(string, data.char_set);
        } else {
            // Fixed width: match exactly 'width' characters
            return matchCharClassFixed(string, data.char_set, data.width);
        }
    }
};

/// Match variable-width character class.
/// Returns the length of the longest prefix where all characters are in the set.
/// Returns null if first character doesn't match (atoms only match non-empty prefixes).
fn matchCharClassVariable(string: []const u8, char_set: CharSet) ?usize {
    var count: usize = 0;
    for (string) |c| {
        if (char_set.contains(c)) {
            count += 1;
        } else {
            break;
        }
    }
    // Atoms only match non-empty prefixes
    return if (count > 0) count else null;
}

/// Match fixed-width character class.
/// Returns width if exactly 'width' characters match, null otherwise.
fn matchCharClassFixed(string: []const u8, char_set: CharSet, width: u32) ?usize {
    if (string.len < width) return null;

    // Check if first 'width' characters are all in the set
    for (string[0..width]) |c| {
        if (!char_set.contains(c)) return null;
    }

    return width;
}

// ============================================================================
// Default Atoms from the FlashProfile Paper (Figure 6)
// ============================================================================

/// Create lowercase letter atom [a-z]
pub fn lower() Atom {
    const chars = "abcdefghijklmnopqrstuvwxyz";
    return .{
        .name = "Lower",
        .static_cost = 9.1,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create uppercase letter atom [A-Z]
pub fn upper() Atom {
    const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    return .{
        .name = "Upper",
        .static_cost = 8.2,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create digit atom [0-9]
pub fn digit() Atom {
    const chars = "0123456789";
    return .{
        .name = "Digit",
        .static_cost = 8.2,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create alphabetic atom [a-zA-Z]
pub fn alpha() Atom {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    return .{
        .name = "Alpha",
        .static_cost = 15.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create alphanumeric atom [a-zA-Z0-9]
pub fn alphaDigit() Atom {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    return .{
        .name = "AlphaDigit",
        .static_cost = 20.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create whitespace atom (space, tab, newline, carriage return, form feed)
pub fn space() Atom {
    const chars = " \t\n\r\x0C";
    return .{
        .name = "Space",
        .static_cost = 5.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create binary digits atom [01]
pub fn bin() Atom {
    const chars = "01";
    return .{
        .name = "Bin",
        .static_cost = 5.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create hexadecimal digits atom [0-9a-fA-F]
pub fn hex() Atom {
    const chars = "0123456789abcdefABCDEF";
    return .{
        .name = "Hex",
        .static_cost = 26.3,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create alphanumeric and whitespace atom [a-zA-Z0-9\s]
pub fn alphaDigitSpace() Atom {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 \t\n\r\x0C";
    return .{
        .name = "AlphaDigitSpace",
        .static_cost = 25.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create dot and dash atom [.-]
pub fn dotDash() Atom {
    const chars = ".-";
    return .{
        .name = "DotDash",
        .static_cost = 3.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create common punctuation atom [.,:?/-]
pub fn punct() Atom {
    const chars = ".,:?/-";
    return .{
        .name = "Punct",
        .static_cost = 10.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create alphabetic and dash atom [a-zA-Z-]
pub fn alphaDash() Atom {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-";
    return .{
        .name = "AlphaDash",
        .static_cost = 18.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create symbol characters atom [-.,://@#$%&*()!~`+=<>?]
pub fn symb() Atom {
    const chars = "-.,://@#$%&*()!~`+=<>?";
    return .{
        .name = "Symb",
        .static_cost = 30.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create alphabetic and whitespace atom [a-zA-Z\s]
pub fn alphaSpace() Atom {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ \t\n\r\x0C";
    return .{
        .name = "AlphaSpace",
        .static_cost = 18.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create Base64 characters atom [a-zA-Z0-9+=]
pub fn base64() Atom {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+=";
    return .{
        .name = "Base64",
        .static_cost = 25.0,
        .data = .{ .char_class = .{
            .char_set = CharSet.fromChars(chars),
            .width = 0,
        } },
    };
}

/// Create any printable ASCII atom (ASCII 32-126)
pub fn any() Atom {
    @setEvalBranchQuota(2000);
    var char_set = CharSet.empty();
    var i: u8 = 32;
    while (i <= 126) : (i += 1) {
        char_set.add(i);
    }
    return .{
        .name = "Any",
        .static_cost = 100.0,
        .data = .{ .char_class = .{
            .char_set = char_set,
            .width = 0,
        } },
    };
}

/// Create a constant string atom.
/// Cost is proportional to 1/length to prefer longer matches.
pub fn constant(name: []const u8, string: []const u8) Atom {
    const len = string.len;
    const cost = 100.0 / @as(f64, @floatFromInt(len));

    return .{
        .name = name,
        .static_cost = cost,
        .data = .{ .constant = .{ .string = string } },
    };
}

/// Create a fixed-width variant of a character class atom.
pub fn withFixedWidth(base_atom: Atom, width: u32) Atom {
    var result = base_atom;
    if (result.data == .char_class) {
        result.data.char_class.width = width;
        // Fixed-width cost is base_cost / width
        result.static_cost = base_atom.static_cost / @as(f64, @floatFromInt(width));
    }
    return result;
}

// ============================================================================
// Tests
// ============================================================================

test "CharSet basic operations" {
    var set = CharSet.empty();
    try std.testing.expect(!set.contains('a'));

    set.add('a');
    try std.testing.expect(set.contains('a'));
    try std.testing.expect(!set.contains('b'));

    set.add('z');
    try std.testing.expect(set.contains('z'));
    try std.testing.expect(set.contains('a'));
}

test "CharSet fromChars" {
    const set = CharSet.fromChars("abc123");
    try std.testing.expect(set.contains('a'));
    try std.testing.expect(set.contains('b'));
    try std.testing.expect(set.contains('c'));
    try std.testing.expect(set.contains('1'));
    try std.testing.expect(set.contains('2'));
    try std.testing.expect(set.contains('3'));
    try std.testing.expect(!set.contains('d'));
    try std.testing.expect(!set.contains('0'));
}

test "Constant atom matching" {
    const pmc = constant("PMC", "PMC");

    try std.testing.expectEqual(@as(?usize, 3), pmc.match("PMC12345"));
    try std.testing.expectEqual(@as(?usize, null), pmc.match("XYZ"));
    try std.testing.expectEqual(@as(?usize, null), pmc.match("PM"));
}

test "Variable-width char class matching" {
    const d = digit();

    try std.testing.expectEqual(@as(?usize, 3), d.match("123abc"));
    try std.testing.expectEqual(@as(?usize, null), d.match("abc123"));
    try std.testing.expectEqual(@as(?usize, 1), d.match("5"));
}

test "Fixed-width char class matching" {
    const digit2 = withFixedWidth(digit(), 2);

    try std.testing.expectEqual(@as(?usize, 2), digit2.match("12345"));
    try std.testing.expectEqual(@as(?usize, null), digit2.match("1abc"));
    try std.testing.expectEqual(@as(?usize, 2), digit2.match("123"));
}

test "Default atoms" {
    const l = lower();
    try std.testing.expectEqual(@as(?usize, 3), l.match("abc123"));
    try std.testing.expectEqual(@as(?usize, null), l.match("ABC"));

    const u = upper();
    try std.testing.expectEqual(@as(?usize, 3), u.match("ABC123"));
    try std.testing.expectEqual(@as(?usize, null), u.match("abc"));

    const a = alpha();
    try std.testing.expectEqual(@as(?usize, 6), a.match("abcDEF123"));

    const s = space();
    try std.testing.expectEqual(@as(?usize, 3), s.match("   abc"));
    try std.testing.expectEqual(@as(?usize, 1), s.match("\ttest"));
}

test "Enhanced atoms - DotDash" {
    const dd = dotDash();
    // DotDash only matches '.' and '-', not digits
    try std.testing.expectEqual(@as(?usize, null), dd.match("2023-01-15"));
    try std.testing.expectEqual(@as(?usize, 2), dd.match("--comment"));
    try std.testing.expectEqual(@as(?usize, 3), dd.match("...more"));
    try std.testing.expectEqual(@as(?usize, null), dd.match("abc"));
    try std.testing.expectEqual(@as(?usize, 1), dd.match("-dash"));
}

test "Enhanced atoms - Symb" {
    const s = symb();
    try std.testing.expectEqual(@as(?usize, 1), s.match("@user"));
    try std.testing.expectEqual(@as(?usize, 1), s.match("#tag"));
    // "://" matches 3 symbol chars (: / /)
    try std.testing.expectEqual(@as(?usize, 3), s.match("://rest"));
    try std.testing.expectEqual(@as(?usize, null), s.match("abc"));
}

test "Enhanced atoms - AlphaSpace" {
    const as = alphaSpace();
    try std.testing.expectEqual(@as(?usize, 11), as.match("Hello World123"));
    // "Hello World  \tMore" has space+space+tab+letters, all match AlphaSpace
    try std.testing.expectEqual(@as(?usize, 18), as.match("Hello World  \tMore"));
    try std.testing.expectEqual(@as(?usize, null), as.match("123"));
}

test "Enhanced atoms - Bin, Hex, Base64" {
    const b = bin();
    try std.testing.expectEqual(@as(?usize, 4), b.match("1010abc"));
    try std.testing.expectEqual(@as(?usize, null), b.match("2"));

    const h = hex();
    // "face123" - all 7 chars are hex digits
    try std.testing.expectEqual(@as(?usize, 7), h.match("face123"));
    try std.testing.expectEqual(@as(?usize, 3), h.match("ABC"));

    const b64 = base64();
    try std.testing.expectEqual(@as(?usize, 5), b64.match("aB3+="));
    try std.testing.expectEqual(@as(?usize, null), b64.match("@"));
}

test "Enhanced atoms - AlphaDigitSpace, AlphaDash, Punct" {
    const ads = alphaDigitSpace();
    try std.testing.expectEqual(@as(?usize, 13), ads.match("Hello World 1"));

    const ad = alphaDash();
    try std.testing.expectEqual(@as(?usize, 10), ad.match("test-case-123"));

    const p = punct();
    try std.testing.expectEqual(@as(?usize, 1), p.match(".txt"));
    // "://" has 3 punct chars (: / /)
    try std.testing.expectEqual(@as(?usize, 3), p.match("://"));
}
